File: SteelPans.Shared/Config/Diagnostics.cs
namespace SteelPans.Shared.Config;

public sealed class ConnectionDetails
{
    public string? Url { get; set; }
    public string? UserAgent { get; set; }
    public string? Platform { get; set; }
    public string? Language { get; set; }

    public ScreenDetails? Screen { get; set; }
    public ViewportDetails? Viewport { get; set; }

    public bool Mobile { get; set; }
    public bool Touch { get; set; }
    public bool Online { get; set; }

    public NetworkConnectionDetails? Connection { get; set; }
}

public sealed class ScreenDetails
{
    public int Width { get; set; }
    public int Height { get; set; }
}

public sealed class ViewportDetails
{
    public int Width { get; set; }
    public int Height { get; set; }
}

public sealed class NetworkConnectionDetails
{
    public string? EffectiveType { get; set; }
    public string? Type { get; set; }

    public double? Downlink { get; set; }
    public int? Rtt { get; set; }

    public bool SaveData { get; set; }
}

File: SteelPans.WebApp/wwwroot/js/diagnostics.js
window.diagnostics = {
    logInitialConnection: () => {
        const nav = navigator;

        const connection =
            nav.connection ||
            nav.mozConnection ||
            nav.webkitConnection;

        const details = {
            url: window.location.href,
            userAgent: nav.userAgent,
            platform: nav.platform,
            language: nav.language,
            screen: {
                width: window.screen.width,
                height: window.screen.height
            },
            viewport: {
                width: window.innerWidth,
                height: window.innerHeight
            },
            mobile: /Android|iPhone|iPad|iPod|Mobile/i.test(nav.userAgent),
            touch: navigator.maxTouchPoints > 0,
            online: navigator.onLine,
            connection: connection
                ? {
                    effectiveType: connection.effectiveType,
                    type: connection.type,
                    downlink: connection.downlink,
                    rtt: connection.rtt,
                    saveData: connection.saveData
                }
                : null
        };

        return details;
    }
};